File: clients/Cpp/model/ServerMessage.cpp
#include "ServerMessage.hpp"
#include <stdexcept>

ServerMessage::GetAction::GetAction() { }
ServerMessage::GetAction::GetAction(PlayerView playerView, bool debugAvailable) : playerView(playerView), debugAvailable(debugAvailable) { }
ServerMessage::GetAction ServerMessage::GetAction::readFrom(InputStream& stream) {
    ServerMessage::GetAction result;
    result.playerView = PlayerView::readFrom(stream);
    result.debugAvailable = stream.readBool();
    return result;
}
void ServerMessage::GetAction::writeTo(OutputStream& stream) const {
    stream.write(TAG);
    playerView.writeTo(stream);
    stream.write(debugAvailable);
}

ServerMessage::Finish::Finish() { }
ServerMessage::Finish ServerMessage::Finish::readFrom(InputStream& stream) {
    ServerMessage::Finish result;
    return result;
}
void ServerMessage::Finish::writeTo(OutputStream& stream) const {
    stream.write(TAG);
}

ServerMessage::DebugUpdate::DebugUpdate() { }
ServerMessage::DebugUpdate::DebugUpdate(PlayerView playerView) : playerView(playerView) { }
ServerMessage::DebugUpdate ServerMessage::DebugUpdate::readFrom(InputStream& stream) {
    ServerMessage::DebugUpdate result;
    result.playerView = PlayerView::readFrom(stream);
    return result;
}
void ServerMessage::DebugUpdate::writeTo(OutputStream& stream) const {
    stream.write(TAG);
    playerView.writeTo(stream);
}
std::shared_ptr<ServerMessage> ServerMessage::readFrom(InputStream& stream) {
    switch (stream.readInt()) {
    case 0:
        return std::shared_ptr<ServerMessage::GetAction>(new ServerMessage::GetAction(ServerMessage::GetAction::readFrom(stream)));
    case 1:
        return std::shared_ptr<ServerMessage::Finish>(new ServerMessage::Finish(ServerMessage::Finish::readFrom(stream)));
    case 2:
        return std::shared_ptr<ServerMessage::DebugUpdate>(new ServerMessage::DebugUpdate(ServerMessage::DebugUpdate::readFrom(stream)));
    default:
        throw std::runtime_error("Unexpected tag value");
    }
};

File: clients/Scala/src/main/scala/model/ServerMessage.scala
package model

import util.StreamUtil

sealed trait ServerMessage {
    def writeTo(stream: java.io.OutputStream)
}
object ServerMessage {
    case class GetAction(playerView: model.PlayerView, debugAvailable: Boolean) extends ServerMessage {
        override def writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, GetAction.TAG)
            playerView.writeTo(stream)
            StreamUtil.writeBoolean(stream, debugAvailable)
        }
    }
    object GetAction {
        val TAG: Int = 0
        def readFrom(stream: java.io.InputStream): GetAction = GetAction(
            model.PlayerView.readFrom(stream)
            ,
            StreamUtil.readBoolean(stream)
            )
    }

    case class Finish() extends ServerMessage {
        override def writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, Finish.TAG)
        }
    }
    object Finish {
        val TAG: Int = 1
        def readFrom(stream: java.io.InputStream): Finish = Finish(
            )
    }

    case class DebugUpdate(playerView: model.PlayerView) extends ServerMessage {
        override def writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, DebugUpdate.TAG)
            playerView.writeTo(stream)
        }
    }
    object DebugUpdate {
        val TAG: Int = 2
        def readFrom(stream: java.io.InputStream): DebugUpdate = DebugUpdate(
            model.PlayerView.readFrom(stream)
            )
    }

    def readFrom(stream: java.io.InputStream): ServerMessage = {
        StreamUtil.readInt(stream) match {
            case GetAction.TAG => GetAction.readFrom(stream)
            case Finish.TAG => Finish.readFrom(stream)
            case DebugUpdate.TAG => DebugUpdate.readFrom(stream)
            case _ => throw new java.io.IOException("Unexpected tag value")
        }
    }
}
